File: rd-cpp/rd_framework_cpp/src/main/base/IProtocol.cpp
//
// Created by jetbrains on 30.07.2018.
//

#include <utility>

#include "IScheduler.h"
#include "IIdentities.h"
#include "IWire.h"
#include "IProtocol.h"

IProtocol::IProtocol(std::shared_ptr<IIdentities> identity, IScheduler *scheduler, std::shared_ptr<IWire> wire) :
        identity(std::move(identity)),
        scheduler(scheduler),
        wire(std::move(wire)),
        context(&serializers) {}

const SerializationCtx &IProtocol::get_serialization_context() const {
    return context;
}

File: rd-net/RdFramework/IProtocol.cs
using System;
using JetBrains.Collections.Viewable;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;

namespace JetBrains.Rd
{
  public interface IProtocol : IRdDynamic
  {
    Lifetime Lifetime { get; }
    SerializationCtx SerializationContext { get; }
    ViewableSet<RdExtBase> OutOfSyncModels { get; }
    string Name { get; }
    ISerializers Serializers { get; }
    IIdentities Identities { get; }
    IScheduler Scheduler { get; }
    IWire Wire { get; }    
    ProtocolContexts Contexts { get; }
    ISignal<ExtCreationInfoEx> ExtCreated { get; }
    
    public T? GetExtension<T>() where T : RdExtBase;
    public T GetOrCreateExtension<T>(Func<T> create) where T : RdExtBase;
  }
}
